csrc/host_ir/executor.cpp

```diff
@@ -195,8 +195,13 @@ void HostIrExecutor::handle(ForLoop* for_loop) {
   auto stop = for_loop->stop()->value().as<int64_t>();
 
   for (auto i = start; i < stop; i += step) {
+    // invalidate i and its consumers before binding
```

Comment:
As mentioned offline, I think this will break for cases like
```
sum = zero tensor
for i: 
  sum += Slice(input, i)
```
because the code would invalidate the accumulator when entering each iteration. 

One possible solution is to force-evaluate top-level expressions in execution order, simply using ExpressionEvaluator as a map.

Reply:
> As mentioned offline, I think this will break for cases like
> 
> ```
> sum = zero tensor
> for i: 
>   sum += Slice(input, i)
> ```
>
> because the code would invalidate the accumulator when entering each iteration.
> 

1. I fail to see how the case you mention is related to the present pr. We do not need such accumulation mechanism, so the current implementation does not face this issue. I suggest to save this concern for a future pr.
2. In general in Fuser, supporting this type of accumulation needs *aliasing*, as can be seen in [this example](https://github.com/NVIDIA/Fuser/pull/2441/files). Indeed, we need a way to differentiate the consumer and the producer in the DAG, which is hidden by using the "`+=`" operator in your example. To make this idea more precise, I imagine in the future a program like
```
sum_in = zero tensor
sum_out = aliasOf(sum_in)
for i: 
  sum_out = add(sum_in, Slice(input, i))
```
This formulation is compatible with the invalidation of the producers of `i` (which are `sum_out`, `Slice(input, i)`, but not `sum_in`), provided we do "alias lookup" for inferring the location of the output buffer, and not "evaluate".

_Incidental note: the above formulation points to the fact that `sum_out` plays two different roles: it represents the result of the compute (obtained through "evaluate"), but it also indicates what buffers stores the results (obtained through aliasing). I don't think there a general treatment of this in Fuser. I don't have a cristal clear idea on how it should be done but in general we should differentiate the definition of a TensorView in the DAG and how to access the buffer associated with it_

> One possible solution is to force-evaluate top-level expressions in execution order, simply using ExpressionEvaluator as a map.

Sorry I don't understand your suggestion. In the current example, there is one single top level expression: the for-loop.

Reply:
I was thinking about something like
```
void HostIrExecutor::handle(ForLoop* for_loop) {
  ...
  for (i = start; i < stop; i += step) {
    ee_.bind(for_loop->index(), i);
    for (Expr* expr : for_loop->body().exprs()) {
      ee_.invalidate(expr's outputs);
      dispatch(expr);
    }
  }
}
```
without traversing def-use chains.

Reply:
> the above formulation points to the fact that sum_out plays two different roles: it represents both the result of the compute and what buffer stores the results.

Yeah, that's a good idea. Real definitions are either an allocation or an ATen (or whatever library we may use in the future) call that implicitly allocates the result tensor.

Reply:
> I was thinking about something like
> 
> ```
> void HostIrExecutor::handle(ForLoop* for_loop) {
>   ...
>   for (i = start; i < stop; i += step) {
>     ee_.bind(for_loop->index(), i);
>     for (Expr* expr : for_loop->body().exprs()) {
>       ee_.invalidate(expr's outputs);
>       dispatch(expr);
>     }
>   }
> }
> ```
> 
> without traversing def-use chains.

I see. Unfortunately, this won't work. The problem is that there might be some intermediate computations between binding the for-loop index `j` and for-loop's body's expressions. More precisely, in our example, the first expression in the for-loop is slicing `tva_` on range `j`, `j+1`, to produce `tva_j`. If we invalidate only `j` and `tva_j`, as suggested in your comment, we won't invalidate `j+1` (which is produced by `add(j, hic->oneVal())`. In reality, there are many of those intermediate computes of indices and predicates.

Besides, doing as you suggest does not help in the situation of buffer reuse in accumulation problems, that you mentioned in the first place, so I fail to see the benefit of it.

_Incidental remark: I want to try switching to `SelectOp` instead of `SliceOp` in the hope it will make the arithmetic simpler which could benefit the semantic but also the performance...._